File: lab3/ProducerConsumerJava/src/com/company/Manager.java
package com.company;

import java.util.ArrayList;
import java.util.concurrent.Semaphore;

public class Manager {
    public final Semaphore access;
    public final Semaphore full;
    public final Semaphore empty;

    public final ArrayList<String> storage = new ArrayList<>();
    private int totalProduced = 0;
    private final int totalItems;

    public Manager(int storageSize, int totalItems) {
        this.totalItems = totalItems;
        access = new Semaphore(1);
        full  = new Semaphore(storageSize);
        empty = new Semaphore(0);
    }

    public synchronized boolean registerProduced() {
        if (totalProduced < totalItems) {
            totalProduced++;
            return true;
        }
        return false;
    }
}

File: lab3/ProducerConsumer/ProducerConsumer/Manager.cs
using System.Collections.Generic;
using System.Threading;

namespace ProducerConsumer
{
    public class Manager
    {
        public Semaphore Access;
        public Semaphore Full;
        public Semaphore Empty;

        public List<string> Storage = new List<string>();

        private int totalProduced = 0;
        private readonly int totalItems;
        private readonly object lockObj = new object();

        public Manager(int storageSize, int totalItems)
        {
            this.totalItems = totalItems;
            Access = new Semaphore(1, 1);
            Full   = new Semaphore(storageSize, storageSize);
            Empty  = new Semaphore(0, storageSize);
        }

        public bool RegisterProduced()
        {
            lock (lockObj)
            {
                if (totalProduced < totalItems)
                {
                    totalProduced++;
                    return true;
                }
                return false;
            }
        }
    }
}